staging/src/k8s.io/apiserver/pkg/registry/rest/validate.go

```diff
@@ -106,3 +111,137 @@ func parseSubresourcePath(subresourcePath string) ([]string, error) {
+func gatherDeclarativeValidationMismatches(imperativeErrs, declarativeErrs field.ErrorList) []string {
+	var mismatchDetails []string
+	// short circuit here to minimize allocs for usual case of 0 validation errors
+	if len(imperativeErrs) == 0 && len(declarativeErrs) == 0 {
+		return mismatchDetails
+	}
+
+	// recommendation based on takeover status
+	recommendation := "This difference should not affect system operation since hand written validation is authoritative."
+	if utilfeature.DefaultFeatureGate.Enabled(features.DeclarativeValidationTakeover) {
+		recommendation = "Consider disabling the DeclarativeValidationTakeover feature gate."
+	}
+
+	matcher := fldtest.ErrorMatcher{}.ByType().ByField().ByOrigin().RequireOriginWhenInvalid()
+	matchedDeclarative := make([]bool, len(declarativeErrs))
+	// Match each "covered" imperative error to a declarative error.
+	for _, iErr := range imperativeErrs {
+		if !iErr.CoveredByDeclarative {
+			continue
+		}
+		foundMatch := false
+		for j, dErr := range declarativeErrs {
+			if !matchedDeclarative[j] && matcher.Matches(iErr, dErr) {
+				matchedDeclarative[j] = true
+				foundMatch = true
+				break
+			}
+		}
+		if !foundMatch {
+			mismatchDetails = append(mismatchDetails,
+				fmt.Sprintf(
+					"Unexpected difference between hand written validation and declarative validation error results, unmatched error(s) found %s. "+
+						"This may indicate an issue with declarative validation. %s",
+					matcher.Render(iErr),
+					recommendation,
+				),
+			)
+		}
+	}
+	// Any remaining unmatched declarative errors are considered "extra".
+	for j, dErr := range declarativeErrs {
+		if !matchedDeclarative[j] {
+			mismatchDetails = append(mismatchDetails,
+				fmt.Sprintf(
+					"Unexpected difference between hand written validation and declarative validation error results, extra error(s) found %s. "+
+						"This may indicate an issue with declarative validation. %s",
+					matcher.Render(dErr),
+					recommendation,
+				),
+			)
+		}
+	}
+	return mismatchDetails
+}
+
+// createDeclarativeValidationPanicHandler returns a function with panic recovery logic
+// that will increment metrics and either log or append errors based on feature gate settings.
+func createDeclarativeValidationPanicHandler(errs *field.ErrorList) func() {
+	return func() {
+		if r := recover(); r != nil {
+			// Increment the panic metric counter
+			validationmetrics.Metrics.EmitDeclarativeValidationPanicMetric()
+
+			const errorFmt = "panic during declarative validation: %v"
+			if utilfeature.DefaultFeatureGate.Enabled(features.DeclarativeValidationTakeover) {
+				// If takeover is enabled, output as a validation error as authorative validator panicked and validation should error
+				*errs = append(*errs, field.InternalError(nil, fmt.Errorf(errorFmt, r)))
+			} else {
+				// if takeover not enabled, log the panic as a warning
+				klog.Warningf(errorFmt, r)
```

Comment:
Just wanted to flag the use of a gate here in our utility functions in apiserver.  We do check flags in the apiserver in other places.  But I can't help but wonder if this would be cleaner if we were able to somehow only return the InternalError in this function and avoid burying this feature gate check.

Reply:
Oh, I see we're using the flags in `gatherDeclarativeValidationMismatches`.  This is probably fine for now.